django/contrib/auth/hashers.py

```diff
@@ -358,7 +384,7 @@ def harden_runtime(self, password, encoded):
         # hardening algorithm.
         pass
 
-    def _decode(self, encoded):
+    def decode(self, encoded):
```

Comment:
This method returns a list and not a dict.

Reply:
Refactored to return a dict. I also used the new decode throughout the argon2 hasher, but skipped the `verify` function. In testing there were two different forms and I don't see how I can cleanly handle both.

Reply:
`v` would be the version of `argon2i` which is present in newer versions of the library. Where do you have problems handling this?

Reply:
The issue is that I don't see a clean way to use decode here. In the argon2 `verify` a split is done
```
algorithm, rest = encoded.split('$', 1)
```
Which results in two cases for the `rest` part and is used directly in
```
return argon2.low_level.verify_secret(
                ('$' + rest).encode('ascii'),
                password.encode(),
                type=argon2.low_level.Type.I,
            )
```
To replace that split call with decode I would need to reconstruct the string with some conditional logic. I don't see a clean way to do that hence I just left the split call in that instance.

Reply:
Ah yes, makes sense, `argon` is to complicated in that matter most likely :/

Reply:
Ya. I certainly could rewrite this bit with decode and I did go through the process. The resulting code was pure in the sense that the only splitting done was in the decode function, but it was also something like 20 extra lines which would probably make no sense until the next reader dug into the commit/PR history.

Reply:
No worries, try to keep the PR small. I will update the Argon hasher myself (updating to a new argon-cffi version) in a separate PR which should (tm) make this code nicer.